benchmark/framework/containerd_utils.go

```diff
@@ -0,0 +1,304 @@
+	image, pullErr := proc.Client.Pull(proc.Context, imageRef, GetRemoteOpts(proc.Context, platform)...)
+	if pullErr != nil {
+		return nil, pullErr
+	}
+	return image, nil
+}
+
+func (proc *ContainerdProcess) CreateContainer(
+	image containerd.Image,
+	opts ...containerd.NewContainerOpts) (containerd.Container, func(), error) {
+	id := fmt.Sprintf("%s-%d", testContainerId, time.Now().UnixNano())
+	opts = append(opts, containerd.WithNewSnapshot(id, image))
+	opts = append(opts, containerd.WithNewSpec(oci.WithImageConfig(image)))
+	container, err := proc.Client.NewContainer(
+		proc.Context,
+		id,
+		opts...)
+	if err != nil {
+		return nil, nil, err
+	}
+	cleanupFunc := func() {
+		err = container.Delete(proc.Context, containerd.WithSnapshotCleanup)
+		if err != nil {
+			fmt.Printf("Error deleting container: %v\n", err)
+		}
+	}
+	return container, cleanupFunc, nil
+}
+
+type TaskDetails struct {
+	task         containerd.Task
+	stdoutReader io.Reader
+	stderrReader io.Reader
+}
+
+func (proc *ContainerdProcess) CreateTask(
+	container containerd.Container) (*TaskDetails, func(), error) {
+	stdoutPipeReader, stdoutPipeWriter := io.Pipe()
+	stderrPipeReader, stderrPipeWriter := io.Pipe()
+	cioCreator := cio.NewCreator(cio.WithStreams(os.Stdin, stdoutPipeWriter, stderrPipeWriter))
+	task, err := container.NewTask(proc.Context, cioCreator)
+	if err != nil {
+		return nil, nil, err
+	}
+	cleanupFunc := func() {
+		stdoutPipeReader.Close()
+		stdoutPipeWriter.Close()
+		stderrPipeReader.Close()
+		stderrPipeWriter.Close()
+		processStatus, err := task.Status(proc.Context)
+		if processStatus.Status != "stopped" {
+			fmt.Printf("Tried to kill task")
+			err = task.Kill(proc.Context, syscall.SIGKILL)
+			if err != nil {
+				fmt.Printf("Error killing task: %v\n", err)
+			}
+		}
+		// stat, err := task.Status(proc.Context)
+		// fmt.Printf("task status before delete %v\n", stat)
+		_, err = task.Delete(proc.Context)
+		if err != nil {
+			fmt.Printf("Error deleting task: %v\n", err)
+		}
+	}
+
+	return &TaskDetails{
+			task,
+			stdoutPipeReader,
+			stderrPipeReader},
+		cleanupFunc,
+		err
+}
+
+func (proc *ContainerdProcess) RunContainerTaskForReadyLine(
+	taskDetails *TaskDetails,
+	readyLine string) (func(), error) {
+	stdoutScanner := bufio.NewScanner(taskDetails.stdoutReader)
+	stderrScanner := bufio.NewScanner(taskDetails.stderrReader)
+
+	exitStatusC, err := taskDetails.task.Wait(proc.Context)
+	if err != nil {
+		return nil, err
+	}
+	resultChannel := make(chan string, 1)
+	go func() {
+		_ = <-exitStatusC
+		resultChannel <- "PROC_EXIT"
+	}()
+	go func() {
+		continueLoop := true
+		for stderrScanner.Scan() && continueLoop {
+			nextLine := stderrScanner.Text()
+			if strings.Contains(nextLine, readyLine) {
+				resultChannel <- "READYLINE_STDERR"
+				continueLoop = false
+			}
+			select {
+			case _ = <-resultChannel:
+				continueLoop = false
+			default:
+				continueLoop = true
+			}
+		}
```

Comment:
I am new to goroutines, so there may be no better way to do this and I wouldn't know. That said, this loop of polling *feels* like unnecessary cpu churn that I would expect to be able to do with a higher level construct of some sort.